File: bench/affine_mix.py
#!/usr/bin/env python3
# benchmark-contract: affine-mix-xs9;seed=123456789;iterations=20000000;mask=0x03ffffffffffffff;xorshift=9
#
# affine_mix — 20M rounds of two chained affine steps over a 58-bit
# state. Python integers are arbitrary precision, so every step masks
# back to 58 bits explicitly to get the fixed-width semantics the other
# four languages have built in.
#
# Contract: print exactly one line — the checksum, masked to 63 bits.
MASK = 0x03FFFFFFFFFFFFFF

state = 123456789
for i in range(20_000_000):
    state = ((state << 3) + i) & MASK
    state ^= state >> 9
    state = ((state << 2) - 3) & MASK

print(state & 0x7FFFFFFFFFFFFFFF)

File: bench/affine_mix.js
// benchmark-contract: affine-mix;seed=123456789;iterations=20000000;mask=0x03ffffffffffffff
//
// affine_mix — 20M rounds of two chained affine steps over a 58-bit
// state.
//
// Representation: BigInt. `state << 3` needs 61 bits and a JS Number is
// only exact to 53, so unlike the other 32-bit kernels in this set
// there is no Number-based path that computes the same value. BigInt is
// what modern JS offers for exact 64-bit-class integers, and that cost
// is part of the answer this row reports.
//
// Contract: print exactly one line — the checksum, masked to 63 bits.
const MASK = 0x03ffffffffffffffn;
const ITERATIONS = 20000000n;

let state = 123456789n;
for (let i = 0n; i < ITERATIONS; i++) {
    state = ((state << 3n) + i) & MASK;
    state = ((state << 2n) - 3n) & MASK;
}

console.log((state & 0x7fffffffffffffffn).toString());
